doc/memos/rdm-draft-alamos-lanzieri-runtime-configuration-architecture.md

```diff
@@ -0,0 +1,447 @@
+Configuration System (RCS) focusing on modularity, reuse of existing
+technologies in RIOT (network stack, storage interface) and standard compliance.
+
+# 2. Architecture
+The proposed RCS architecture is formed by a
+[configuration manager](4-configuration-manager) and the
+[RIOT Registry](3-the-riot-registry) and its sub-components:
+one or more [Registry Handlers](4-registry-handlers) and one or more
+[storage facilities](5-storage-facilities). 
+
+A RIOT Application may interact with the Configuration Manager in order to
+modify access control rules or enable different communication interfaces.
+Also, it may interact with the RIOT Registry directly if it's
+required to load/store a persistent configuration.
+
+![](./files/rdm-draft-alamos-lanzieri-runtime-configuration-architecture/architecture.svg "Runtime Configuration Architecture")
+
+# 3. The RIOT Registry
+The RIOT Registry is a module for interacting with 
+**persistent key-value configurations**. It's heavily inspired in the
+[Mynewt Config subsystem](https://mynewt.apache.org/latest/os/modules/config/config.html)
+
+The Registry interacts with RIOT components via Registry Handlers, and with
+non-volatile storage devices via Storage Facilities. This way the Registry is
+independent to the functionality of the module or storage device.
+
+<img src="./files/rdm-draft-alamos-lanzieri-runtime-configuration-architecture/riot-registry-diagram.svg" />
+
+The API of the RIOT Registry allows to:
+- Register a RH to expose a configuration group in the RIOT Registry.
+- Register source and destination SF.
+- Get or set configuration parameters for a given configuration group.
+- Commit changes (transactionally apply configurations)
+- Load and store configuration parameters from a persistent storage device
+- Export configuration parameters (e.g copy to a buffer, print, etc)
+
+Any mechanism of security (access control, encryption of configurations) is NOT
+directly in the scope of the Registry but in the Configuration Manager and the
+specific implementation of the RH and SF.
+
+See [3.3 RIOT Registry Usage Flow](#3.3. RIOT Registry Usage Flow) for more
+information.
+
+## 3.1. Registry handlers
+A RH represents a configuration group in the RIOT Registry. A RIOT
+module requires to implement and register a Registry Handler in order to expose
+its configurations to the Registry API.
+
+A RH is defined by a name and a series of handlers for
+interacting with the configuration parameters of the configuration group.
+These handlers are:
+- `set`: Sets a value to a configuration parameter. This handler MUST take care
+  of any logic to apply the value (e.g. data validation) or cache it until the
+commit handler is called
+- `get`: Gets the current value of a configuration parameter
+- `commit`: To be called when configuration parameters have been loaded from
+  storage. It could be use for special needed logic on applying configuration
```

Comment:
this is stated in the Usage Flow section:

> As one could expect, registry_load will call the SF load handler with registry_set_value as callback. In the a similar way, registry_store will navigate through all RH and call their export function with the SF store handler as callback.

Reply:
Yep ok. What I meant was that saying

 "`commit`: To be called when configuration parameters have been loaded from storage"
Is a bit confusing, because `set` can also be called when configuration parameters have been loaded from storage. It would be less confusing to omit this sentence entirely here. Soemthing like:
`commit`: does the same as `set`, but also allows specially needed logic to be executed on applying configuration

Reply:
> "commit: To be called when configuration parameters have been loaded from storage

I will remove this sentence

> commit: does the same as set, but also allows specially needed logic to be executed on applying configuration

Just for reference, `commit` does NOT do exactly the same as `set`. In fact, commit doesn't receive any keys or values.

The `set` handler only "sets" the value and the configuration MAY take effect. However, the only way to make sure all configurations take effect is calling `commit` after a batch of `set` calls

Reply:
Does that mean that "set" will change the value of the variable, but it may not take effect because maybe some other logic needs to be done for that to happen?

Reply:
yes

Reply:
Ah ok I get it now. Can I suggest:
`commit`: To be called once configuration parameters have been `set`, in order to apply any further logic required to make them effective (e.g. handling dependencies).

Reply:
yes sure. Much better, thanks!